pkg/controller/validation/ippool.go

```diff
@@ -0,0 +1,189 @@
+
+package validation
+
+import (
+	"fmt"
+	"net/netip"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	crdv1beta1 "antrea.io/antrea/pkg/apis/crd/v1beta1"
+	utilip "antrea.io/antrea/pkg/util/ip"
+)
+
+// GetIPRangeSet returns a set of string representations of IP ranges
+func GetIPRangeSet(ipRanges []crdv1beta1.IPRange) sets.Set[string] {
+	set := sets.New[string]()
+	for _, ipRange := range ipRanges {
+		ipRangeStr := ipRange.CIDR
+		if ipRangeStr == "" {
+			ipRangeStr = fmt.Sprintf("%s-%s", ipRange.Start, ipRange.End)
+		}
+		set.Insert(ipRangeStr)
+	}
+	return set
+}
+
+// parseIPRangeCIDR parses a CIDR string into a netip.Prefix
+func parseIPRangeCIDR(cidrStr string) (netip.Prefix, error) {
+	cidr, err := netip.ParsePrefix(cidrStr)
+	if err != nil {
+		return cidr, fmt.Errorf("invalid cidr %s", cidrStr)
+	}
+	return cidr.Masked(), nil
+}
+
+// parseIPRangeStartEnd parses start and end IP addresses
+func parseIPRangeStartEnd(startStr, endStr string) (netip.Addr, netip.Addr, error) {
+	start, err := netip.ParseAddr(startStr)
+	if err != nil {
+		return start, netip.Addr{}, fmt.Errorf("invalid start ip address %s", startStr)
+	}
+
+	end, err := netip.ParseAddr(endStr)
+	if err != nil {
+		return start, end, fmt.Errorf("invalid end ip address %s", endStr)
+	}
+	return start, end, nil
+}
+
+// validateIPRange validates an IP range specification
+func validateIPRange(ipRange crdv1beta1.IPRange) error {
+	start, end, err := parseIPRangeStartEnd(ipRange.Start, ipRange.End)
+	if err != nil {
+		return err
+	}
+
+	if start.Is4() != end.Is4() {
+		return fmt.Errorf("range start %s and range end %s should belong to same family",
+			ipRange.Start, ipRange.End)
+	}
+
+	if start.Compare(end) > 0 {
+		return fmt.Errorf("range start %s should not be greater than range end %s",
+			ipRange.Start, ipRange.End)
+	}
+	return nil
+}
+
+// ValidateIPRangesAndSubnetInfo validates IP ranges and SubnetInfo
+func ValidateIPRangesAndSubnetInfo(subnetInfo *crdv1beta1.SubnetInfo, ipRanges []crdv1beta1.IPRange) ([]NormalizedIPRange, error) {
+	var subnet *netip.Prefix
+	if subnetInfo != nil {
+		gatewayAddr, err := netip.ParseAddr(subnetInfo.Gateway)
+		if err != nil {
+			return nil, fmt.Errorf("invalid gateway address %s", subnetInfo.Gateway)
+		}
+
+		if gatewayAddr.Is4() {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 32 {
+				return nil, fmt.Errorf("invalid prefixLength %d", subnetInfo.PrefixLength)
+			}
+		} else {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
+				return nil, fmt.Errorf("invalid prefixLength %d", subnetInfo.PrefixLength)
+			}
+		}
+		prefix := netip.PrefixFrom(gatewayAddr, int(subnetInfo.PrefixLength)).Masked()
+		subnet = &prefix
+	}
+
+	currentRanges := make([]NormalizedIPRange, 0, len(ipRanges))
+	for _, ipRange := range ipRanges {
+		cur, err := normalizeRange(ipRange, "")
+		if err != nil {
+			return nil, err
+		}
+		start, end := cur.Start, cur.End
+		key := cur.Origin
```

Comment:
nit: I actually think it's cleared if you don't define these variables and just use `cur` directly

Reply:
done